### DISCUSSION TITLE
Prevent rofi from hanging window manager untill it reads all options from stdin

### DISCUSSION THREAD

**Discussion by atk91** [davatorium/rofi#1486] [(Possible) Issues] [answered]
Hi everyone,
I use i3 window manager, maybe this will also arise in other ones.

If you execute this
`{sleep 3; echo 'a\nb'} | rofi -no-steal-focus -dmenu`
your desktop will be completely irresponsible for 3 seconds. Not sure what's going on, but it looks like rofi still does some kind of focus stealing and waits until it can display the window, but for 3 seconds it obviously cannot.

I'm also not sure whether it's a bug or not, so I added (Possible) Issues category.

Originally I faced that when `ssh user@host -- "cmd_producing_options" | rofi -dmenu | do-something-with-selected-option`. If your ssh key requires a password - it hangs your i3wm completely. Only thing left is to restart X.

I've found a workaround - use `-normal-window` switch. Is there anything better?

**Comment by DaveDavenport** [comment, marked as answer]
Yes the `-async-pre-read 0` option should do the trick.

**Reply by atk91** [reply]
Thanks, it indeed does.
But maybe it would be better if rofi had an option to read whole stdin without trying to do anything with X. One may suppose that `-sync` option does so, since man page says
```
       -sync

       Force  rofi  mode  to first read all data from stdin before showing the
       selection window.
```
But it still blocks your wm while reading stdin. Is it an intended behaviour?

**Reply by DaveDavenport** [reply]
dmenu is not rofi's main purpose, its designed to be as quick as possible in its normal mode. 
Here it grabs the keyboard as quick as possible to not loose keypresses.
(I guess this is also what people expect for dmenu mode)
